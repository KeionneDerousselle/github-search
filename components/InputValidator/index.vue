<template>
  <ValidationProvider
    v-slot="{ errors, failedRules, aria, classes, validate, reset, valid, invalid, changed, touched, untouched, pristine, dirty, pending, required, validated, passed, failed }"
    :vid="vid"
    :tag="tag"
    :skip-if-empty="skipIfEmpty"
    :rules="rules"
    :name="name || label"
    :mode="mode"
    :immediate="immediate"
    :disabled="disabled"
    :debounce="debounce"
    :custom-messages="customMessages"
    :bails="bails"
    class="input">
    <label
      v-if="label"
      :for="labelFor"
      class="input__label sr-only">
      {{ label }}
    </label>

    <slot
      :errors="errors"
      :failedRules="failedRules"
      :aria="aria"
      :classes="classes"
      :validate="validate"
      :reset="reset"
      :valid="valid"
      :invalid="invalid"
      :changed="changed"
      :touched="touched"
      :untouched="untouched"
      :pristine="pristine"
      :dirty="dirty"
      :pending="pending"
      :required="required"
      :validated="validated"
      :passed="passed"
      :failed="failed" />

    <slot
      v-if="!errors || errors.length < 1"
      name="hint"
      :errors="errors"
      :failedRules="failedRules"
      :aria="aria"
      :classes="classes"
      :validate="validate"
      :reset="reset"
      :valid="valid"
      :invalid="invalid"
      :changed="changed"
      :touched="touched"
      :untouched="untouched"
      :pristine="pristine"
      :dirty="dirty"
      :pending="pending"
      :required="required"
      :validated="validated"
      :passed="passed"
      :failed="failed">
      <div
        v-if="hint"
        class="input__hint-container">
        <span class="input__hint text-black-500">{{ hint }}</span>
      </div>
    </slot>

    <slot
      v-if="showErrors && errors && errors.length > 0"
      name="errors"
      :errors="errors"
      :failedRules="failedRules"
      :classes="classes"
      :valid="valid"
      :invalid="invalid"
      :changed="changed"
      :touched="touched"
      :untouched="untouched"
      :pristine="pristine"
      :dirty="dirty"
      :pending="pending"
      :required="required"
      :validated="validated"
      :passed="passed"
      :failed="failed">
      <div class="input__errors">
        <span
          v-if="bails"
          class="input__error text-error-400">
          {{ errors[0] }}
        </span>
        <span
          v-for="error in errors"
          v-else
          :key="`${uuid}-error-${error}`"
          class="input__error text-error-400">
          {{ error }}
        </span>
      </div>
    </slot>
  </ValidationProvider>
</template>
<script>
import { ValidationProvider } from 'vee-validate'
export default {
  name: 'InputValidator',

  components: {
    ValidationProvider
  },

  inheritAttrs: false,

  props: {
    bails: {
      type: Boolean,
      default: true
    },

    customMessages: {
      type: Object,
      default: () => ({})
    },

    debounce: {
      type: Number,
      default: 0
    },

    disabled: {
      type: Boolean,
      default: false
    },

    immediate: {
      type: Boolean,
      default: false
    },

    mode: {
      type: String,
      default: 'eager',
      validator: value =>
        [
          'aggressive',
          'passive',
          'lazy',
          'eager'
        ].includes(value)
    },

    name: {
      type: String,
      default: ''
    },

    rules: {
      type: [ String, Object ],
      default: null
    },

    skipIfEmpty: {
      type: Boolean,
      default: true
    },

    tag: {
      type: String,
      default: 'div'
    },

    vid: {
      type: String,
      default: ''
    },

    labelFor: {
      type: String,
      default: undefined
    },

    label: {
      type: String,
      default: undefined
    },

    hint: {
      type: String,
      default: ''
    },

    showErrors: {
      type: Boolean,
      default: true
    }
  }
}
</script>
<style lang="scss">
.input__errors,
.input__hint-container {
  margin-top: 0.4rem;
}

.input__error {
  font-size: 1.2rem;
  font-weight: 500;
  display: block;
  font-family: inherit;
}

.input__hint {
  display: block;
  font-size: 1.2rem;
  font-family: inherit;
}
</style>
