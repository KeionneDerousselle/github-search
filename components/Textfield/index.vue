<template>
  <div class="attrs-fix">
    <InputValidator
      v-slot="{errors}"
      :vid="vid"
      :tag="tag"
      :skip-if-empty="skipIfEmpty"
      :rules="rules"
      :name="veeValidateName"
      :label="label"
      :label-for="$attrs.id"
      :mode="mode"
      :immediate="immediate"
      :disabled="validationDisabled"
      :debounce="debounce"
      :custom-messages="customMessages"
      :bails="bails"
      :hint="hint"
      class="textfield-container">
      <div
        tabindex="0"
        :class="['textfield', 'rounded-lg', calcContainerClasses(errors, isFocused)]"
        @focus="handleContainerFocused"
        @click="handleContainerFocused"
        @blur="blur">
        <span
          v-if="$slots.left"
          class="textfield__left">
          <slot name="left" />
        </span>

        <input
          :ref="`textfield-input-${$attrs.id}`"
          v-model="text"
          :class="['textfield__input', {'textfield__input--readonly': readonly}, calcInputClasses(errors, isFocused)]"
          :type="$attrs.type || 'text'"
          :placeholder="placeholder || label"
          :readonly="readonly"
          :disabled="disabled"
          :name="name"
          v-bind="$attrs"
          @input="handleEvent"
          @blur="handleInputBlurred"
          @focus="focus">

        <span
          v-if="$slots.right"
          class="textfield__right">
          <slot name="right" />
        </span>
      </div>
    </InputValidator>
  </div>
</template>
<script>
import InputValidator from '@/components/InputValidator'

export default {
  name: 'Textfield',

  components: {
    InputValidator
  },

  inheritAttrs: false,

  props: {
    bails: {
      type: Boolean,
      default: true
    },

    customMessages: {
      type: Object,
      default: () => ({})
    },

    debounce: {
      type: Number,
      default: 0
    },

    validationDisabled: {
      type: Boolean,
      default: false
    },

    immediate: {
      type: Boolean,
      default: false
    },

    mode: {
      type: String,
      default: 'eager',
      validator: value =>
        [
          'aggressive',
          'passive',
          'lazy',
          'eager'
        ].includes(value)
    },

    veeValidateName: {
      type: String,
      default: ''
    },

    name: {
      type: String,
      default: ''
    },

    rules: {
      type: [ String, Object ],
      default: null
    },

    skipIfEmpty: {
      type: Boolean,
      default: true
    },

    tag: {
      type: String,
      default: 'div'
    },

    vid: {
      type: String,
      default: ''
    },

    label: {
      type: String,
      required: true
    },

    calcContainerClasses: {
      type: Function,
      default: () => ([])
    },

    calcInputClasses: {
      type: Function,
      default: () => ([])
    },

    value: {
      type: String,
      default: ''
    },

    hint: {
      type: String,
      default: ''
    },

    placeholder: {
      type: String,
      default: ''
    },

    disabled: {
      type: Boolean,
      default: false
    },

    readonly: {
      type: Boolean,
      default: false
    }
  },

  data: () => ({
    text: '',
    isFocused: false
  }),

  watch: {
    value(newVal) {
      this.text = newVal
    }
  },

  methods: {
    handleEvent(event) {
      this.$emit(event.type, event.target.value)
    },

    focus() {
      this.isFocused = true
    },

    handleContainerFocused() {
      const inputRef = `textfield-input-${this.$attrs.id}`

      this.focus()
      this.$refs[inputRef].focus()
    },

    blur() {
      this.isFocused = false
    },

    handleInputBlurred(event) {
      this.handleEvent(event)
      this.blur()
    }
  }
}
</script>
<style lang="scss">
.attrs-fix,
.textfield-container,
.textfield,
.textfield__input {
  @apply w-full;
}

.textfield {
  @apply relative p-3 inline-flex overflow-hidden items-center text-base p-4 font-normal;

  transition: all 0.5s;

  &__left {
    @apply mr-2.5 inline-flex;
  }

  &__right {
    @apply ml-2.5 inline-flex;
  }
}

.textfield__input {
  @apply appearance-none m-0 border-0 outline-none flex-1;

  background: inherit;
  color: inherit;
  font-family: inherit;
  font-weight: inherit;

  &:-webkit-autofill,
  &:-webkit-autofill:hover,
  &:-webkit-autofill:focus,
  &:-webkit-autofill:active,
  &:-internal-autofill-selected {
    color: inherit;
    font-family: inherit;
    font-weight: inherit;
    background-clip: text;
  }

  &::placeholder {
    @apply font-normal;

    font-family: inherit;
  }

  &::-ms-clear,
  &::-ms-reveal {
    @apply hidden w-0 h-0;
  }

  &::-webkit-search-decoration,
  &::-webkit-search-cancel-button,
  &::-webkit-search-results-button,
  &::-webkit-search-results-decoration {
    @apply appearance-none;
  }

  &:disabled,
  &--readonly {
    cursor: default;
  }
}
</style>
