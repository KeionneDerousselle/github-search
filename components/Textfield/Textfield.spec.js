import flushPromises from 'flush-promises'
import { extend } from 'vee-validate'
import { required } from 'vee-validate/dist/rules'
import InputValidator from '../InputValidator'
import Textfield from '.'

describe('TextField', () => {
  let wrapper
  let inputClasses
  const label = 'My Label'

  beforeAll(() => {
    extend('required', required)
  })

  describe('when no props are provided', () => {
    beforeAll(() => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })
      inputClasses = wrapper.get('.textfield__input').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not add the readonly attribute to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().readonly).toBeFalsy()
    })

    it('should not add the readonly classes to the text field', () => {
      expect(inputClasses).not.toContain('textfield__input--readonly')
    })

    it('should not add the disabled attribute to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().disabled).toBeFalsy()
    })

    // it('should not add the disabled classes to the text field', () => {
    // })

    it('should add the label as the placeholder to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().placeholder).toBe(label)
    })

    it('should not add the name to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().name).toBe('')
    })
  })

  describe('when a placeholder is provided', () => {
    let placeholder

    beforeAll(() => {
      placeholder = 'My Placeholder'

      wrapper = mountPreMocked(Textfield, {
        propsData: {
          placeholder,
          label
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should add the placeholder to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().placeholder).toBe(placeholder)
    })
  })

  describe('when a placeholder is not provided', () => {
    beforeAll(() => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should add the placeholder to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().placeholder).toBe(label)
    })
  })

  describe('when a name is provided', () => {
    let name

    beforeAll(() => {
      name = 'inputName'

      wrapper = mountPreMocked(Textfield, {
        propsData: {
          name,
          label
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should add the name to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().name).toBe(name)
    })
  })

  describe('when a name is not provided', () => {
    beforeAll(() => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not add the name to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().name).toBe('')
    })
  })

  describe('when the text field is disabled', () => {
    beforeAll(() => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          disabled: true,
          label
        }
      })

      inputClasses = wrapper.get('.textfield__input').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should add the disabled attribute to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().disabled).toBeTruthy()
    })

    // it('should add the disabled classes to the text field', () => {
    // })
  })

  describe('when the text field is enabled', () => {
    beforeAll(() => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          disabled: false,
          label
        }
      })

      inputClasses = wrapper.get('.textfield__input').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not add the disabled attribute to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().disabled).toBeFalsy()
    })

    // it('should not add the disabled classes to the text field', () => {
    // })
  })

  describe('when the text field is readonly', () => {
    beforeAll(() => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          readonly: true,
          label
        }
      })
      inputClasses = wrapper.get('.textfield__input').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should add the readonly attribute to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().readonly).toBeTruthy()
    })

    it('should add the readonly classes to the text field', () => {
      expect(inputClasses).toContain('textfield__input--readonly')
    })
  })

  describe('when the text field is not readonly', () => {
    beforeAll(() => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          readonly: false,
          label
        }
      })
      inputClasses = wrapper.get('.textfield__input').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not add the readonly attribute to the text field', () => {
      expect(wrapper.get('.textfield__input').attributes().readonly).toBeFalsy()
    })

    it('should not add the readonly classes to the text field', () => {
      expect(inputClasses).not.toContain('textfield__input--readonly')
    })
  })

  describe('when input classes are provided', () => {
    let customClass
    let calcInputClasses

    beforeAll(() => {
      customClass = 'my-custom-class'
      calcInputClasses = jest.fn(() => customClass)

      wrapper = mountPreMocked(Textfield, {
        propsData: {
          calcInputClasses,
          label
        }
      })

      inputClasses = wrapper.get('.textfield__input').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should call the custom input classes function', () => {
      expect(calcInputClasses).toHaveBeenCalled()
    })

    it('should add custom css classes to the text field input tag', () => {
      expect(inputClasses).toContain(customClass)
    })
  })

  describe('when input classes are not provided', () => {
    let customClass
    let calcInputClasses

    beforeAll(() => {
      customClass = 'my-custom-class'
      calcInputClasses = jest.fn(() => customClass)

      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })

      inputClasses = wrapper.get('.textfield__input').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should not call the custom input classes function', () => {
      expect(calcInputClasses).not.toHaveBeenCalled()
    })

    it('should not add custom css classes to the text field input tag', () => {
      expect(inputClasses).not.toContain(customClass)
    })
  })

  describe('when container classes are provided', () => {
    let customClass
    let calcContainerClasses

    beforeAll(() => {
      customClass = 'my-custom-class'
      calcContainerClasses = jest.fn(() => customClass)

      wrapper = mountPreMocked(Textfield, {
        propsData: {
          calcContainerClasses,
          label
        }
      })

      inputClasses = wrapper.get('.textfield').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should call the custom container classes function', () => {
      expect(calcContainerClasses).toHaveBeenCalled()
    })

    it('should add custom css classes to the text field container', () => {
      expect(inputClasses).toContain(customClass)
    })
  })

  describe('when container classes are not provided', () => {
    let customClass
    let calcContainerClasses

    beforeAll(() => {
      customClass = 'my-custom-class'
      calcContainerClasses = jest.fn(() => customClass)

      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })

      inputClasses = wrapper.get('.textfield').classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should not call the custom container classes function', () => {
      expect(calcContainerClasses).not.toHaveBeenCalled()
    })

    it('should not add custom css classes to the text field container', () => {
      expect(inputClasses).not.toContain(customClass)
    })
  })

  describe('when validation errors are present', () => {
    beforeAll(async () => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          rules: 'required',
          label
        }
      })

      const input = wrapper.get('.textfield__input')

      input.setValue('')
      await flushPromises()
      await wrapper.vm.$nextTick()

      input.trigger('blur')
      await flushPromises()
      await wrapper.vm.$nextTick()

      inputClasses = input.classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should add the error classes to the text field', () => {
    })
  })

  describe('when validation errors are not present', () => {
    beforeAll(async () => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          rules: 'required',
          label
        }
      })

      const input = wrapper.get('.textfield__input')

      input.setValue('abc')
      await flushPromises()
      await wrapper.vm.$nextTick()

      input.trigger('blur')
      await flushPromises()
      await wrapper.vm.$nextTick()

      inputClasses = input.classes()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not add the error classes to the text field', () => {
    })
  })

  describe('when the input event is emitted from the input', () => {
    beforeAll(async () => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })

      const input = wrapper.get('.textfield__input')

      input.setValue('a')
      await flushPromises()
      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should emit the input event from the text field', () => {
      expect(wrapper.emitted().input[0][0]).toBe('a')
    })
  })

  describe('when the focus event is emitted from the input', () => {
    beforeAll(async () => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })

      const input = wrapper.get('.textfield__input')

      input.trigger('focus')
      await flushPromises()
      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should set the \'isFocused\' to be true', () => {
      expect(wrapper.vm.isFocused).toBe(true)
    })
  })

  describe('when the blur event is emitted from the input', () => {
    beforeAll(async () => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })

      const input = wrapper.get('.textfield__input')

      input.setValue('a')
      await flushPromises()
      await wrapper.vm.$nextTick()

      input.trigger('blur')
      await flushPromises()
      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should emit the blur event from the text field', () => {
      expect(wrapper.emitted().blur[0][0]).toBe('a')
    })

    it('should set the \'isFocused\' flag to false', () => {
      expect(wrapper.vm.isFocused).toBe(false)
    })
  })

  describe('when the focus event is emitted from the container element', () => {
    let inputFocusSpy

    beforeAll(async () => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })
      await wrapper.vm.$nextTick()

      inputFocusSpy = jest.spyOn(wrapper.vm.$refs[`textfield-input-${wrapper.vm.$attrs.id}`], 'focus')

      const input = wrapper.get('.textfield')

      input.trigger('focus')
      await flushPromises()
      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      inputFocusSpy.mockRestore()
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should focus the inner input element', () => {
      expect(inputFocusSpy).toHaveBeenCalled()
    })

    it('should set the \'isFocused\' to be true', () => {
      expect(wrapper.vm.isFocused).toBe(true)
    })
  })

  describe('when the blur event is emitted from the container element', () => {
    beforeAll(async () => {
      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })

      const input = wrapper.get('.textfield')

      input.trigger('focus')
      await flushPromises()
      await wrapper.vm.$nextTick()

      input.trigger('blur')
      await flushPromises()
      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should set the \'isFocused\' flag to false', () => {
      expect(wrapper.vm.isFocused).toBe(false)
    })
  })

  describe('when a new value is passed into the text field', () => {
    let value

    beforeAll(async () => {
      value = 'abc'

      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })
      await wrapper.vm.$nextTick()

      wrapper.setProps({
        value
      })
      await wrapper.vm.$nextTick()
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should set the text data to the passed in value', () => {
      expect(wrapper.vm.text).toBe(value)
    })
  })

  describe('when rendering the text field', () => {
    let inputValidator

    beforeAll(() => {
      wrapper = mountPreMocked(Textfield, {
        attrs: {
          id: 'my-text-field'
        },

        propsData: {
          vid: 'myTextFieldValidationId',
          tag: 'p',
          skipIfEmpty: true,
          rules: 'required',
          veeValidateName: 'Validation Field Name',
          label,
          mode: 'aggressive',
          immediate: true,
          validationDisabled: false,
          debounce: 300,
          bails: false,
          customMessages: { required: '{_field_} is missing.' },
          hint: 'This is a hint.'
        }
      })
      inputValidator = wrapper.findComponent(InputValidator)
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should pass the vid prop to the input validator', () => {
      expect(inputValidator.vm.vid).toBe(wrapper.vm.vid)
    })

    it('should pass the tag prop to the input validator', () => {
      expect(inputValidator.vm.tag).toBe(wrapper.vm.tag)
    })

    it('should pass the skip if empty prop to the input validator', () => {
      expect(inputValidator.vm.skipIfEmpty).toBe(wrapper.vm.skipIfEmpty)
    })

    it('should pass the rules prop to the input validator', () => {
      expect(inputValidator.vm.rules).toBe(wrapper.vm.rules)
    })

    it('should pass the vee validate name prop to the input validator', () => {
      expect(inputValidator.vm.name).toBe(wrapper.vm.veeValidateName)
    })

    it('should pass the label prop to the input validator', () => {
      expect(inputValidator.vm.label).toBe(wrapper.vm.label)
    })

    it('should pass the id attribute as the label-for prop to input validator', () => {
      expect(inputValidator.vm.labelFor).toBe(wrapper.vm.$attrs.id)
    })

    it('should pass the mode prop to the input validator', () => {
      expect(inputValidator.vm.mode).toBe(wrapper.vm.mode)
    })

    it('should pass the immediate prop to the input validator', () => {
      expect(inputValidator.vm.immediate).toBe(wrapper.vm.immediate)
    })

    it('should pass the validation disabled prop to the input validator', () => {
      expect(inputValidator.vm.disabled).toBe(wrapper.vm.validationDisabled)
    })

    it('should pass the debounce prop to the input validator', () => {
      expect(inputValidator.vm.debounce).toBe(wrapper.vm.debounce)
    })

    it('should pass the custom messages prop to the input validator', () => {
      expect(inputValidator.vm.customMessages).toBe(wrapper.vm.customMessages)
    })

    it('should pass the bails prop to the input validator', () => {
      expect(inputValidator.vm.bails).toBe(wrapper.vm.bails)
    })

    it('should pass the hint prop to the input validator', () => {
      expect(inputValidator.vm.hint).toBe(wrapper.vm.hint)
    })
  })

  describe('when a right slot is provided', () => {
    let rightSlotId

    beforeAll(() => {
      rightSlotId = 'right-slot'

      wrapper = mountPreMocked(Textfield, {
        slots: {
          right: `<span id="${rightSlotId}">R</span>`
        },

        propsData: {
          label
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render the right content slot', () => {
      expect(wrapper.get(`.textfield__right #${rightSlotId}`).element).toBeVisible()
    })

    it('should not render the left slot', () => {
      expect(() => wrapper.get('.textfield__left')).toThrow()
    })
  })

  describe('when a right slot is not provided', () => {
    let rightSlotId

    beforeAll(() => {
      rightSlotId = 'right-slot'

      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not render the right content slot', () => {
      expect(() => wrapper.get(`.textfield__right #${rightSlotId}`)).toThrow()
    })

    it('should not render the left slot', () => {
      expect(() => wrapper.get('.textfield__left')).toThrow()
    })
  })

  describe('when a left slot is provided', () => {
    let leftSlotId

    beforeAll(() => {
      leftSlotId = 'left-slot'

      wrapper = mountPreMocked(Textfield, {
        slots: {
          left: `<span id="${leftSlotId}">L</span>`
        },

        propsData: {
          label
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render the left content slot', () => {
      expect(wrapper.get(`.textfield__left #${leftSlotId}`).element).toBeVisible()
    })

    it('should not render the right slot', () => {
      expect(() => wrapper.get('.textfield__right')).toThrow()
    })
  })

  describe('when a left slot is not provided', () => {
    let leftSlotId

    beforeAll(() => {
      leftSlotId = 'left-slot'

      wrapper = mountPreMocked(Textfield, {
        propsData: {
          label
        }
      })
    })

    afterAll(() => {
      wrapper.destroy()
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should not render the left content slot', () => {
      expect(() => wrapper.get(`.textfield__left #${leftSlotId}`)).toThrow()
    })

    it('should not render the right slot', () => {
      expect(() => wrapper.get('.textfield__right')).toThrow()
    })
  })
})
