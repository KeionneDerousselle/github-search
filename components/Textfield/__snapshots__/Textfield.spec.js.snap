// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField when a name is not provided should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when a name is provided should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="inputName" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when a placeholder is not provided should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when a placeholder is provided should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Placeholder" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when no props are provided should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when the focus event is emitted from the container element should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when the input event is emitted from the input should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when the text field is disabled should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" disabled="disabled" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when the text field is enabled should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when the text field is not readonly should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when the text field is readonly should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" readonly="readonly" name="" type="text" class="textfield__input textfield__input--readonly">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when validation errors are not present should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`TextField when validation errors are present should render as expected 1`] = `
<div class="attrs-fix">
  <div class="input textfield-container"><label class="input__label sr-only">
      My Label
    </label>
    <div tabindex="0" class="textfield rounded-lg">
      <!----> <input placeholder="My Label" name="" type="text" class="textfield__input">
    </div>
    <!---->
    <div class="input__errors"><span class="input__error text-error-400">
        My Label is not valid.
      </span></div>
  </div>
</div>
`;
